numpyro/distributions/distribution.py

```diff
@@ -444,28 +444,41 @@ def has_rsample(self):
         return self.base_dist.has_rsample
 
     def _sample(self, sample_fn, key, sample_shape=()):
-        interstitial_dims = tuple(self._interstitial_sizes.keys())
-        event_dim = len(self.event_shape)
-        interstitial_dims = tuple(i - event_dim for i in interstitial_dims)
         interstitial_sizes = tuple(self._interstitial_sizes.values())
         expanded_sizes = tuple(self._expanded_sizes.values())
         batch_shape = expanded_sizes + interstitial_sizes
-        samples = sample_fn(key, sample_shape=sample_shape + batch_shape)
-        interstitial_idx = len(sample_shape) + len(expanded_sizes)
-        interstitial_sample_dims = tuple(range(interstitial_idx, interstitial_idx + len(interstitial_sizes)))
-        for dim1, dim2 in zip(interstitial_dims, interstitial_sample_dims):
-            samples = jnp.swapaxes(samples, dim1, dim2)
-        return samples.reshape(sample_shape + self.batch_shape + self.event_shape)
+        samples, intermediates = sample_fn(key, sample_shape=sample_shape + batch_shape)
+
+        def reshape_sample(x):
+            """ Reshapes samples and intermediates to ensure that the output
+                shape is correct: This implicitly replaces the interstitial dims
+                of size 1 in the original batch_shape of base_dist with those
+                in the expanded dims. While it somewhat 'shuffles' over batch
+                dimensions, we don't care because they are considered independent."""
```

Comment:
Makes sense to me. The change looks great but let me think a bit more to see if I miss something.